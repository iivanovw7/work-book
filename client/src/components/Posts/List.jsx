import PropTypes from 'prop-types';
import React, { useEffect, useState, useRef } from 'react';
import moment from 'moment';
import styled from 'styled-components';
import { debounce } from 'debounce';
import ListElementContainer from './ListElement/Container';
import { gridConfig, stylesConfig } from '../../config';
import * as utils from '../../utils';
/* eslint no-underscore-dangle: 0 */

const { appBarHeight, topBarHeight } = stylesConfig;

const Container = styled.div`
  display: flex;
  flex-direction: column;
  
  @media screen and (max-width: ${gridConfig.container.md}rem) {
    padding-bottom: ${props => (props.hasMore ? `calc(2em + ${stylesConfig.appBarHeight}px)` : '0em')};
  }
  
  @media screen and (min-width: ${gridConfig.container.md}rem) {
    padding-bottom: ${props => (props.hasMore ? '5em' : '0em')};
  }
  
`;

const PostsList = (props) => {
  const {
    data, history, locale, atBottomHandler, hasMore, loading, loader
  } = props;
  const { posts } = data.getPosts;
  const [listHeight, setListHeight] = useState(0);
  const [listOffsetTop, setListOffsetTop] = useState(0);
  const listRef = useRef({});

  function setListSize() {
    if (listRef.current) {
      setListHeight(listRef.current.clientHeight);
      setListOffsetTop(listRef.current.offsetTop);
    }
  }

  const debouncedScrollHandler = debounce(() => {
    const windowScroll = window.scrollY + window.innerHeight;
    const listScroll = listHeight + listOffsetTop + appBarHeight + topBarHeight;
    const reachedBottom = (listScroll - windowScroll) <= 100;

    if (reachedBottom) {
      return atBottomHandler();
    }
  }, 200);

  useEffect(() => {
    setListSize();
  }, [debouncedScrollHandler]);

  function handleClick(id) {
    history.push(`/posts/${id}`);
  }

  utils.useEventListener('scroll', event => debouncedScrollHandler(event), window);

  return (
    <Container ref={listRef} hasMore={hasMore}>
      {posts.map((post) => {
        moment.locale(locale.slice(0, -1));
        const date = post.created.toString();
        const est = utils.calculateReadingTime(post.text.length, locale);

        return (
          <ListElementContainer
            key={post._id}
            post={post}
            date={date}
            est={est}
            handleClick={handleClick}
            {...props}
          />
        );
      })}
      {loading && (
        loader
      )}
    </Container>
  );
};

PostsList.propTypes = {
  atBottomHandler: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  hasMore: PropTypes.bool.isRequired,
  history: PropTypes.object.isRequired,
  loader: PropTypes.object.isRequired,
  loading: PropTypes.bool.isRequired,
  locale: PropTypes.string.isRequired
};

export default PostsList;
