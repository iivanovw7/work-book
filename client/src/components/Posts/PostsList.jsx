import PropTypes from 'prop-types';
import React from 'react';
import moment from 'moment';
import styled from 'styled-components';
import { opacify } from 'polished';
import { fadeInTop } from '../../styles';
import { gridConfig } from '../../config';
import Button from '../../elements/UI/Button';
import Number from '../../elements/UI/Number';
import TextLink from '../../elements/UI/TextLink';
import TagsCloud from '../Tags/TagsCloud';
import * as utils from '../../utils';
/* eslint no-underscore-dangle: 0 */

const PostWrapper = styled.div`
	@media screen and (max-width: ${gridConfig.container.md}rem) {
 		flex-direction: column;
  	>div:first-of-type, >div:last-of-type {
  	width: 100%;
  	}
  }
  
  @media screen and (min-width: ${gridConfig.container.md}rem) {
  	flex-direction: row;
  	>div:first-of-type {
  		width: 70%;
  	}
  	
  	>div:last-of-type {
  		width: 30%;
  		display: flex;
  		flex-direction: column;
  		justify-content: space-between;
  	}
  }
  
  ${fadeInTop};
  display: flex;
	flex-wrap: no-wrap;
	justify-content: space-between;
	user-select: none;
	margin: 0.5em 0 0.5em 0;
	padding: 0.7em;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
	
	>div:first-of-type {
		flex-wrap: nowrap;
    overflow-wrap: break-word;
	}
	
	>div:last-of-type {
		text-align: right;
		align-items: flex-end;
	}
`;

const TagsContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: ${props => props.justifyContent};
  margin-top: 1em;
`;

const PostsList = (props) => {
  const {
    data, history, theme, locale, text
  } = props;

  function handleClick(id) {
    history.push(`/posts/${id}`);
  }

  return (data.getPosts.map((post) => {
    moment.locale(locale.slice(0, -1));
    const date = post.created.toString();
    const est = utils.calculateReadingTime(post.text.length, locale);

    return (
      <PostWrapper key={post._id}>
        <div>
          <h2>
            <TextLink
              theme={theme}
              text={post.title}
              variant="secondary"
              link="/"
              handleClick={(e) => {
                e.preventDefault();
                handleClick(post._id);
              }}
            />
          </h2>
          <h4>{post.subject}</h4>
          <div className="mobile">
            <TagsContainer justifyContent="flex-start">
              <TagsCloud
                data={data}
                history={history}
                postTags
                tags={post.tags[0]}
                margin="0.2em 0.2em 0.5em 0em"
                {...props}
              />
            </TagsContainer>
          </div>
          <Number
            variant="primary"
            value={est.minutes}
            label={est.ending}
            theme={theme}
          />
        </div>
        <div className="desktop">
          <div>
            <h3 className="dates">
              {moment(date, 'x')
                .format('MMMM DD, YYYY')}
            </h3>
            <TagsContainer justifyContent="flex-end">
              <TagsCloud
                data={data}
                history={history}
                postTags
                tags={post.tags[0]}
                margin="0.2em"
                {...props}
              />
            </TagsContainer>
          </div>
          <Button
            variant="primary"
            text={text.navigation.readMore[locale]}
            theme={theme}
            maxWidth="10em"
            handleClick={(e) => {
              e.preventDefault();
              handleClick(post._id);
            }}
          />
        </div>
      </PostWrapper>
    );
  }));
};

export default PostsList;

PostsList.propTypes = {
  data: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  locale: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired
};
