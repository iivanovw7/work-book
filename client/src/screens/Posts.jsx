import PropTypes from 'prop-types';
import React from 'react';
import { withRouter } from 'react-router-dom';
import useStoreon from 'storeon/react';
import styled, { ThemeProvider } from 'styled-components';
import QueryPosts from '../graphql/QueryPosts';
import MobileNavBar from '../components/Navigation/MobileNavbar';
import SideBar from '../components/Navigation/SideBar';
import TopBar from '../components/Navigation/TopBar';
import { wrapper, content, aside } from '../styles';
import { backgroundColor, textColor } from '../theme';
/* eslint react/require-default-props: 0 */
/* eslint no-underscore-dangle: 0 */

const StyledSection = styled.section`
  color: ${textColor};
  background-color: ${backgroundColor};
  min-width: 100%;
  box-sizing: content-box;
`;

const StyledWrapper = styled.section`
  ${wrapper}
`;

const StyledSideBar = styled.aside`
  ${aside}
`;

const StyledContainer = styled.div`
  ${content}
`;

const Posts = (props) => {
  const { location } = props;
  const { theme, locale, localizedText } = useStoreon('theme', 'locale', 'localizedText');

  return (
    <ThemeProvider theme={{ mode: theme }}>
      <StyledSection className="container">
        <section>
          <TopBar locale={locale} text={localizedText} {...props} />
        </section>
        <StyledWrapper>
          <StyledSideBar>
            <SideBar
              theme={theme}
              location={location}
              locale={locale}
              text={localizedText}
              onlyMobile={false}
              {...props}
            />
          </StyledSideBar>
          <StyledContainer>
            <QueryPosts locale={locale} text={localizedText} theme={theme} {...props} />
          </StyledContainer>
        </StyledWrapper>
        <section className="mobile">
          <MobileNavBar theme={theme} locale={locale} text={localizedText} {...props} />
        </section>
      </StyledSection>
    </ThemeProvider>
  );
};

Posts.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  user: PropTypes.any,
  userAccess: PropTypes.bool.isRequired
};

export default withRouter(Posts);
