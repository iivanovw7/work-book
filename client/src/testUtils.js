import { createMemoryHistory } from 'history';
import * as localized from './assets/locales.json';

// Mock browser history
export const history = createMemoryHistory('/dashboard');

// Grid config object for testing
export const mappedSizes = {
	xs: 12,
	sm: 12,
	md: 12,
	lg: 11
};

export const localizedText = {
	locales: localized.locales,
	topBar: localized.topBar,
	login: localized.login,
	navigation: localized.navigation,
	pageNotFound: localized.pageNotFound
};
