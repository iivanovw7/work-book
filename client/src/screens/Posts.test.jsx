import React from 'react';
import 'cross-fetch/polyfill';
import 'jsdom-global/register';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import * as testUtils from '../testUtils';
import Posts from './Posts';

jest.mock('../config/apiURL', () => () => {
	return '/api';
});

jest.mock('storeon/react', () => () => {
	return {
		locale: 'eng',
		theme: 'dark'
	};
});
jest.mock('../components/MobileNavBar/MobileNavbar', () => () => (
	<div id="MobileNavbar" />
));
jest.mock('../components/TopBar/TopBar', () => () => (
	<div id="TopBar" />
));
jest.mock('../components/SideBar/SideBar', () => () => (
	<div id="SideBar" />
));
jest.mock('../queryBlocks/QueryPosts', () => () => (
	<div id="QueryPosts" />
));

describe('Testing POSTS screen: ', () => {

	afterAll(() => {
		jest.clearAllMocks();
	});

	const props = {
		localizedText: testUtils.localizedText,
		location: {},
		user: {},
		userAccess: true
	};

	const Composition = () => (
		<MemoryRouter>
			<Posts {...props} />
		</MemoryRouter>
	);

	it('Should render properly with props ', () => {
		const component = mount(<Composition />);
		expect(component.find('Posts').length).toEqual(1);
		expect(component.find('#MobileNavbar').length).toEqual(1);
		expect(component.find('#TopBar').length).toEqual(1);
		expect(component.find('#SideBar').length).toEqual(1);
	});

	it('Should render and match snapshot ', () => {
		const component = mount(<Composition />);
		expect(component).toMatchSnapshot();
	});
});
