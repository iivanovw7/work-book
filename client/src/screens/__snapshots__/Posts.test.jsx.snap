// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing POSTS screen:  Should render and match snapshot  1`] = `ReactWrapper {}`;
